crates/namada/src/ledger/native_vp/masp.rs

```diff
@@ -304,110 +296,327 @@ where
-                    Address::Internal(Masp),
-                ),
-                Ordering::Greater => (
-                    pre_masp_balance - post_masp_balance,
-                    Address::Internal(Masp),
-                    counterpart,
-                ),
-            };
-
-        Ok(TransparentTransferData {
-            source,
-            target,
-            token,
-            amount,
-        })
+            result
+                .decoder
+                .insert(address_hash, counterpart.to_address_ref().clone());
+            *result.pre.entry(address_hash).or_insert(ValueSum::zero()) +=
+                ValueSum::from_pair(token.clone(), pre_balance);
+            *result.post.entry(address_hash).or_insert(ValueSum::zero()) +=
+                ValueSum::from_pair(token.clone(), post_balance);
+        }
+
+        Ok(result)
     }
 }
 
 // Make a map to help recognize asset types lacking an epoch
 fn unepoched_tokens(
     token: &Address,
     denom: token::Denomination,
-) -> Result<HashMap<AssetType, (Address, token::Denomination, MaspDigitPos)>> {
-    let mut unepoched_tokens = HashMap::new();
+    tokens: &mut BTreeMap<
+        AssetType,
+        (Address, token::Denomination, MaspDigitPos),
+    >,
+) -> Result<()> {
     for digit in MaspDigitPos::iter() {
         let asset_type = encode_asset_type(token.clone(), denom, digit, None)
             .wrap_err("unable to create asset type")?;
-        unepoched_tokens.insert(asset_type, (token.clone(), denom, digit));
+        tokens.insert(asset_type, (token.clone(), denom, digit));
+    }
+    Ok(())
+}
+
+// Handle transparent input
+fn validate_transparent_input<A: Authorization>(
+    vin: &TxIn<A>,
+    changed_balances: &mut ChangedBalances,
+    transparent_tx_pool: &mut I128Sum,
+    epoch: Epoch,
+    conversion_state: &ConversionState,
+    signers: &mut BTreeSet<TransparentAddress>,
+) -> Result<()> {
+    // A decrease in the balance of an account needs to be
+    // authorized by the account of this transparent input
+    signers.insert(vin.address);
+    // Non-masp sources add to the transparent tx pool
+    *transparent_tx_pool = transparent_tx_pool
+        .checked_add(
+            &I128Sum::from_nonnegative(vin.asset_type, vin.value as i128)
+                .ok()
+                .ok_or_err_msg("invalid value or asset type for amount")?,
+        )
+        .ok_or_err_msg("Overflow in input sum")?;
+
+    let bal_ref = changed_balances
+        .pre
+        .entry(vin.address)
+        .or_insert(ValueSum::zero());
+
+    match conversion_state.assets.get(&vin.asset_type) {
+        // Note how the asset's epoch must be equal to the present: users
+        // must never be allowed to backdate transparent inputs to a
+        // transaction for they would then be able to claim rewards while
+        // locking their assets for negligible time periods.
+        Some(((address, _asset_denom, digit), asset_epoch, _, _))
```

Comment:
this tuple is super complex. IMO it would be better to extract some struct with named fields

Reply:
Thanks. I've made a structure with named fields to represent this nested tuple in #3264 .